### PR TITLE
More composite updates for integration

### PR DESCRIPTION
**Pull request by spencerhance**
/assign bowei

blocking https://github.com/kubernetes/ingress-gce/pull/788

### CODE REVIEW

File: pkg/composite/utils.go [hunk 1]
```diff
@@ -126,18 +126,20 @@ func IsRegionalUrlMap(um *UrlMap) (bool, error) {
...
 	if err != nil {
-		return false, fmt.Errorf("error parsing self-link for url-map %s: %v", selfLink, err)
+		return false, err
 	}
+	return scope == meta.Regional, nil
+}
 
-	if resourceID.Key.Region != "" {
-		return true, nil
+func ParseScope(selfLink string) (meta.KeyType, error) {
```
Comment (bowei) [inline pkg/composite/utils.go]: unit test?

File: pkg/composite/composite.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ func SetSslCertificateForTargetHttpsProxy(gceCloud *gce.Cloud, key *meta.Key, ta
 
 	// Set name in case it is not present in the key
 	key.Name = targetHttpsProxy.Name
-	klog.V(3).Infof("setting URLMap for TargetHttpsProxy %v", targetHttpsProxy.Name)
+	klog.V(3).Infof("setting SslCertificate for TargetHttpsProxy %v", targetHttpsProxy.Name)
```
Comment (bowei) [inline pkg/composite/composite.go]: Can we fix the comment, it's not clear what resource the key refers to.

klog.V(3).Infof("SetSslCertificate for TargetHttpsProxy %v", key)

File: pkg/composite/utils.go (hunk 1 above)
Comment (spencerhance) [reply pkg/composite/utils.go]: Done

File: pkg/composite/composite.go (hunk 2 above)
Comment (spencerhance) [reply pkg/composite/composite.go]: Done

File: pkg/composite/utils.go [hunk 3]
```diff
@@ -126,18 +126,20 @@ func IsRegionalUrlMap(um *UrlMap) (bool, error) {
...
+	resourceID, err := cloud.ParseResourceURL(selfLink)
+	if err != nil {
+		var empty meta.KeyType
+		return empty, fmt.Errorf("error parsing self-link for url-map %s: %v", selfLink, err)
```
Comment (bowei) [inline pkg/composite/utils.go]: fix this log message (url-map)

File: pkg/composite/utils.go [hunk 4]
```diff
@@ -126,18 +126,20 @@ func IsRegionalUrlMap(um *UrlMap) (bool, error) {
...
 	}
+	return scope == meta.Regional, nil
+}
 
-	if resourceID.Key.Region != "" {
-		return true, nil
+func ParseScope(selfLink string) (meta.KeyType, error) {
+	resourceID, err := cloud.ParseResourceURL(selfLink)
+	if err != nil {
+		var empty meta.KeyType
```
Comment (bowei) [inline pkg/composite/utils.go]: this is not necessary

return meta.KeyType{}, fmt.Errorf(...)

File: pkg/composite/utils_test.go [hunk 5]
```diff
@@ -114,3 +114,51 @@ func TestIsRegionalResource(t *testing.T) {
...
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			result, err := ParseScope(tc.selfLink)
+
+			if tc.wantErr {
+				if err == nil {
+					t.Fatalf("err = nil, want non-nil")
+				} else {
```
Comment (bowei) [inline pkg/composite/utils_test.go]: you don't need this in an else {}

File: pkg/composite/utils.go (hunk 1 above)
Comment (bowei) [inline pkg/composite/utils.go]: probably better name is ScopeFromSelfLink

We are not really parsing a Scope string

File: pkg/composite/utils.go (hunk 4 above)
Comment (spencerhance) [reply pkg/composite/utils.go]: That actually won't compile since KeyType is a string

File: pkg/composite/utils.go (hunk 3 above)
Comment (spencerhance) [reply pkg/composite/utils.go]: Done

File: pkg/composite/utils_test.go (hunk 5 above)
Comment (spencerhance) [reply pkg/composite/utils_test.go]: Done

File: pkg/composite/utils.go (hunk 1 above)
Comment (spencerhance) [reply pkg/composite/utils.go]: Done

File: pkg/composite/utils.go (hunk 4 above)
Comment (spencerhance) [reply pkg/composite/utils.go]: replaced it with meta.KeyType("") and that works

File: pkg/composite/utils_test.go [hunk 6]
```diff
@@ -114,3 +114,50 @@ func TestIsRegionalResource(t *testing.T) {
...
 }
+
+func TestParseScope(t *testing.T) {
+	testCases := []struct {
+		desc     string
+		selfLink string
+		expected meta.KeyType
+		wantErr  bool
+	}{
+		{
```
Comment (bowei) [inline pkg/composite/utils_test.go]: add a testcase for empty string

File: pkg/composite/utils_test.go [hunk 7]
```diff
@@ -114,3 +114,50 @@ func TestIsRegionalResource(t *testing.T) {
 		})
 	}
 }
+
+func TestParseScope(t *testing.T) {
+	testCases := []struct {
```
Comment (bowei) [inline pkg/composite/utils_test.go]: t.Parallel()

**Comment by bowei** [comment]
/approve

ping when done and I will lgtm

**Comment by spencerhance** [comment]
@bowei should be good to go now

**Comment by spencerhance** [comment]
@bowei friendly ping

**Pull request merged**